horovod/run/run_func.py

```diff
@@ -0,0 +1,199 @@
+
+    :param fn: Function to run.
+    :param num_proc: Number of Horovod processes.
+    :param host: To specify the list of host names as well as the
+                 number of available slots on each host for
+                 training processes using the following format:
+                 <hostname>:<number of slots>,... .
+                 E.g., host1:2,host2:4,host3:1 indicates that 2 processes can run on
+                 host1, 4 processes on host2, and 1 process on host3.
+                 If None, launch all mpi processes on local node.
+    :param ssh_port: SSH port on all the hosts.
+                     If None, use default SSH port.
+    :param disable_cache: If the flag is not set, horovodrun will perform
+                          the initialization checks only once every 60
+                          minutes -- if the checks successfully pass.
+                          Otherwise, all the checks will run every time
+                          horovodrun is called."
+    :param start_timeout: Horovodrun has to perform all the checks and
+                          start the processes before the specified
+                          timeout. The default value is 30 seconds.
+                          Alternatively, The environment variable
+                          HOROVOD_START_TIMEOUT can also be used to
+                          specify the initialization timeout.
+    :param env: Environment dictionary to use in Horovod run.
+                If None, use `os.environ`.
+    :param verbose: If this flag is set, extra messages will be printed.
+
+    :return: A list contains returned result from each MPI process. The ith value of the list
+             corresponds to the returned result of rank i process.
+    """
+
+    wdir = tempfile.mkdtemp()
+    if verbose:
+        print("horovod.run_func working dir is " + wdir)
+
+    if ssh_port:
+        ssh_port_opt = "-p " + str(ssh_port)
+        scp_port_opt = "-P " + str(ssh_port)
+    else:
+        ssh_port_opt = ""
+        scp_port_opt = ""
+
+    def wrapped_proc_fn(rank=0):
+        return_value = fn()
+
+        result_file = "{prefix}.{rank}".format(prefix=_PICKLED_RESULT_PREFIX, rank=rank)
+        with open(result_file, 'wb') as fp:
+            try:
+                cloudpickle.dump(return_value, fp)
+            except Exception as e:
+                raise RuntimeError("Caught an excpetion while pickling "
+                                   "return value: {}".format(type(e)), e)
+
+    pickled_proc_fn_str = cloudpickle.dumps(wrapped_proc_fn)
+    local_pickled_proc_fn_path = os.path.join(wdir, _LOCAL_PICKLED_PROC_FN_FILENAME)
+    with open(local_pickled_proc_fn_path, 'wb') as f:
+        f.write(pickled_proc_fn_str)
+
+    local_launcher_path = os.path.join(wdir, _LOCAL_PROC_LAUNCHER_FILENAME)
+    with open(local_launcher_path, 'w') as f:
+        f.write(textwrap.dedent("""
+        set -e
+        cd {wdir}
+        rank=${{OMPI_COMM_WORLD_RANK:-0}}
```

Comment:
I can remove this. We can get the rank id in mpi process code.

Reply:
Thought again. Remove this is troublesome. It requires the arg `fn` return "rank", because I don't know which framework in `fn` user will use, e.g., If tensorflow, get rank by `horovod.tensorflow.rank()`, if pytorch, get rank by `horovod.pytorch.rank()`.